nx/lib/nx.ex

```diff
@@ -1120,6 +1120,91 @@ defmodule Nx do
     backend.eye(%T{type: type, shape: shape, names: names}, backend_options)
   end
 
+  @doc """
+  Extract a diagonal or create a diagonal tensor.
+
+  Extracts a diagonal if given a 2d tensor, or constructs a tensor given a 1d one.
+
+  ## Examples
+
+  When given a 1d tensor:
+
+      iex> Nx.iota({4}) |> Nx.diag
+      #Nx.Tensor<
+        s64[4][4]
+        [
+          [0, 0, 0, 0],
+          [0, 1, 0, 0],
+          [0, 0, 2, 0],
+          [0, 0, 0, 3]
+        ]
+      >
+
+  Given a 2d tensor:
+
+      iex> Nx.iota({3, 3}) |> Nx.diag()
+      #Nx.Tensor<
+        s64[3]
+        [0, 4, 8]
+      >
+
+  And if given a 2d tensor along with an offset:
+
+      iex> Nx.iota({3, 3}) |> Nx.diag(offset: 1)
+      #Nx.Tensor<
+        s64[2]
+        [1, 5]
+      >
+
+      iex> Nx.iota({3, 3}) |> Nx.diag(offset: -1)
+      #Nx.Tensor<
+        s64[2]
+        [3, 7]
+      >
+
+  ## Options
+
+    * `:offset` - offset used for extracting the diagonal.
+      Use offset > 0 for diagonals above the main diagonal,
+      and offset < 0 for diagonals below the main diagonal.
+
+  """
+  @doc type: :creation
+  def diag(tensor, opts \\ [])
+
+  def diag(tensor = %Nx.Tensor{shape: {len, breadth}}, opts) do
+    opts = keyword!(opts, [:offset])
+
+    offset = opts[:offset] || 0
+
+    indices =
+      case offset do
+        i when i >= 0 ->
+          Enum.zip_with(0..(len - 1), i..(breadth - 1), fn x, y -> [x, y] end)
+
+        i when i < 0 ->
+          Enum.zip_with(-i..(len - 1), 0..(breadth - 1), fn x, y -> [x, y] end)
+      end
+
+    indices =
+      indices
+      |> Nx.tensor()
+
+    Nx.gather(tensor, indices)
+  end
+
+  def diag(tensor = %Nx.Tensor{shape: {len}}, _opts) do
+    identity =
+      {len, len}
+      |> Nx.eye()
+
+    Nx.multiply(tensor, identity)
```

Comment:
```suggestion
    Nx.select(identity, tensor, Nx.broadcast(0, tensor))
```

@josevalim I'm not sure which is better for backends like EXLA and Torchx... WDYT?

Reply:
@polvalente I would go with multiply, but the performance difference on EXLA seems to be negligible. Here's a benchmark in JAX:

```python
%%timeit -n 25
jax.jit(lambda identity, tensor: jax.lax.select(identity, tensor, jax.lax.broadcast(0., tensor.shape)))(identity, tensor)
```

```
25 loops, best of 5: 25.5 ms per loop
```

```python
%%timeit -n 25
jax.jit(lambda identity, tensor: jax.lax.mul(identity, tensor))(identity, tensor)
```

```
25 loops, best of 5: 24.2 ms per loop
```

Reply:
Since in EXLA the difference is negligible, let's go with multiply, which will tend to be more memory-efficient for non-compiled backends